### PR TITLE
Don't use DefaultInterpolatedStringHandler in expression trees

### PR DESCRIPTION
**Pull request by 333fred**
Introduces a binder flag to track whether we are currently in an expression tree or not. Fixes https://github.com/dotnet/roslyn/issues/55114.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 1]
```diff
@@ -10316,6 +10316,32 @@ public void DisallowedInExpressionTrees()
...
+        public void AsStringInExpressionTrees_01()
+        {
+            var code = @"
+using System;
+using System.Linq.Expressions;
+
+Expression<Func<string, string>> e = o => $""{o.Length}"";";
+
+            var comp = CreateCompilation(new[] { code, GetInterpolatedStringHandlerDefinition(includeSpanOverloads: false, useDefaultParameters: false, useBoolReturns: false) });
+            comp.VerifyEmitDiagnostics();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10329]: Can we verify the nodes in the expression tree?

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -565,12 +565,13 @@ private BoundExpression CreateAnonymousFunctionConversion(SyntaxNode syntax, Bou
...
                 {
-                    if (destination.IsNonGenericExpressionType())
+                    bool isExpression = destination.IsNonGenericExpressionType();
+                    if (isExpression)
                     {
                         delegateType = Compilation.GetWellKnownType(WellKnownType.System_Linq_Expressions_Expression_T).Construct(delegateType);
                         delegateType.AddUseSiteInfo(ref useSiteInfo);
                     }
-                    boundLambda = unboundLambda.Bind(delegateType);
+                    boundLambda = unboundLambda.Bind(delegateType, isExpression);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:574]: >isExpression [](http://example.com/codeflow?start=67&length=12)

nit: consider naming argument

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 3]
```diff
@@ -10316,6 +10316,32 @@ public void DisallowedInExpressionTrees()
...
         }
 
+        [Fact, WorkItem(55114, "https://github.com/dotnet/roslyn/issues/55114")]
+        public void AsStringInExpressionTrees_01()
+        {
+            var code = @"
+using System;
+using System.Linq.Expressions;
+
+Expression<Func<string, string>> e = o => $""{o.Length}"";";
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10326]: The code suggests that there's also a non-generic `Expression` scenario (see `IsGenericOrNonGenericExpressionType`). Do we need to cover that as well?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs [hunk 4]
```diff
@@ -440,7 +440,7 @@ internal abstract class UnboundLambdaState
         [PerformanceSensitive(
             "https://github.com/dotnet/roslyn/issues/23582",
             Constraint = "Avoid " + nameof(ConcurrentDictionary<NamedTypeSymbol, BoundLambda>) + " which has a large default size, but this cache is normally small.")]
```
Comment (chsienki) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:442]: Nit: update comment to match new signature.

**Comment by 333fred** [comment]
@jcouv @chsienki addressed feedback, thanks. Please take another look.

**Review by chsienki — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 5]
```diff
@@ -10326,20 +10326,237 @@ public void AsStringInExpressionTrees_01()
...
+  IL_0075:  ldc.i4.0
+  IL_0076:  ldloc.0
+  IL_0077:  stelem.ref
+  IL_0078:  call       ""System.Linq.Expressions.Expression<System.Func<string, string>> System.Linq.Expressions.Expression.Lambda<System.Func<string, string>>(System.Linq.Expressions.Expression, params System.Linq.Expressions.ParameterExpression[])""
+  IL_007d:  call       ""System.Linq.Expressions.ParameterExpression[] System.Array.Empty<System.Linq.Expressions.ParameterExpression>()""
+  IL_0082:  call       ""System.Linq.Expressions.Expression<System.Func<System.Func<string, string>>> System.Linq.Expressions.Expression.Lambda<System.Func<System.Func<string, string>>>(System.Linq.Expressions.Expression, params System.Linq.Expressions.ParameterExpression[])""
+  IL_0087:  pop
+  IL_0088:  ret
+}
+");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:10559]: nit: I had a different kind of test in mind (less noisy), possibly using `ExpressionTestLibrary`, but I guess this works too. Sorry, I should have been specific.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Pull request merged**